### PR TITLE
[第8組]Refactor/llm client

### PR DESCRIPTION
**Pull request by hamster0717**
### Purpose
upgrade the readability and functionality of the llm client module by refactoring  code about ai_client, implementing OOP

### Changes
add src/llm_client directory

fixed news/router, news/service, to enable llm_client using in them

### CODE REVIEW

File: = convert_news_to_dict(udn_crawler.parse(news.url))
Comment (mlcp29th-hi) [inline = convert_news_to_dict(udn_crawler.parse(news.url)):1]: What is this? 👀🤔

File: .github/workflows/ci.yaml [hunk 1]
```diff
@@ -37,19 +38,19 @@ jobs:
...
 
-      # - name: Run unit tests
-      #   env:
-      #     OPENAI_API_KEY: ${{ secrets.OPENAI_API_KEY }}
-      #   run: |
-      #     cd backend
-      #     pytest tests/unit --cov=src --cov-report=xml --disable-warnings
+      - name: Run unit tests
+        env:
+          OPENAI_API_KEY: ${{ secrets.OPENAI_API_KEY }}
```
Comment (mlcp29th-hi) [inline .github/workflows/ci.yaml:43]: The environment variable name is different from your project codebase.

File: backend/tests/integration/__pycache__/test_news_endpoint.cpython-312-pytest-8.2.2.pyc
Comment (mlcp29th-hi) [inline backend/tests/integration/__pycache__/test_news_endpoint.cpython-312-pytest-8.2.2.pyc:1]: This file can be excluded from the git repository by adding it to .gitignore file.

File: backend/src/llm_client/base.py [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    system_content: str = Field(...)
+    user_content: str = Field(...)
+
+    @property
+    def to_dict(self):
+        value = [
+            {"role": "system", "content": f"{self.system_content}"},
+            {"role": "user", "content": f"{self.user_content}"},
+        ]
+        return value
```
Comment (mlcp29th-hi) [inline backend/src/llm_client/base.py:18]: 1. Noticing that you pass `...` to `Field`, I'd suggest removing `Field` since it's not required to define the attributes of BaseModel subclass.
2. The `to_dict()` method, with the `@property` decorator, can be accessed as an attribute. I'd recommend naming it as a noun for clarity
3. The `to_dict()` method returns a `list`, not a `dict`.
4. All of the instances `MessageInterface` in you project are created by passing `system_content` and `user_content` directly (e.g., in src/llm_client/llm_client.py, lines 39 to 42):

```py
return self._generate(MessageInterface(
             system_content="你是一個關鍵字提取機器人，用戶將會輸入一段文字，表示其希望看見的新聞內容，請提取出用戶希望看見的關鍵字，請截取最重要的關鍵字即可，避免出現「新聞」、「資訊」等混淆搜尋引擎的字詞。(僅須回答關鍵字，若有多個關鍵字，請以空格分隔)",
             user_content=prompt
         ))
```

If that's the case, inheriting from pydantic.BaseModel seems a bit redundant, and we can simply define it generally.

```suggestion
class MessageInterface:
    def __init__(self, system_content: str, user_content: str):
        self.system_content = system_content
        self.user_content = user_content

    @property
    def llm_messages() -> list:
        ...
```

File: backend/src/llm_client/base.py [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+    
+
+class LLMClientBase(metaclass=abc.ABCMeta):
+    """
+    Abstract defining the structure for an LLM client.
+    """
+    openai_client: OpenAI | None
```
Comment (mlcp29th-hi) [inline backend/src/llm_client/base.py:25]: Creating a base class means it can be inherited to build any specific kind of LLM client, I recommend avoiding tying the property name or type to any specific LLM service in the base class.

```suggestion
class LLMClientBase(metaclass=abc.ABCMeta):
    """
    Abstract defining the structure for an LLM client.
    """
    client: Any
```

File: backend/src/llm_client/openai_client.py [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+from .base import LLMClientBase, MessageInterface
+from .exceptions import LLMClientInitializeException, EvaluationFailure
+from .config import get_ai_config
+ai_config=get_ai_config()
+class LLMClient(LLMClientBase):
+    def __init__(self, _api_key: str = ai_config.OPEN_AI_KEY):
+        try:
+            self.openai_client = OpenAI(api_key=_api_key)
+        except APIError as error:
+            raise LLMClientInitializeException(f"Failed to initialize the LLM client due to an error: {error}")
```
Comment (mlcp29th-hi) [inline backend/src/llm_client/openai_client.py:14]: 1. This class implements the abstract methods inherited from LLMClientBase with OpenAI, I'd recommend naming it as `OpenAIClient`.

2. It seems like you don't have to name the initialization parameter `_api_key` with the single leading underscore.

3. No matter what the `_api_key` is, the initialization of `OpenAI` never raises an exception, so the try-except statement seems a bit redundant.

File: backend/src/llm_client/openai_client.py [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+                )
+                return completion.choices[0].message.content
+            except APIError as error:
+                print(f"[OpenAI] An error occurred: {error}")
+                return ""
+            except RateLimitError as error:
+                print(f"[OpenAI] Rate limit exceeded: {error}")
+                return ""
+        else:
+            return ""
```
Comment (mlcp29th-hi) [inline backend/src/llm_client/openai_client.py:31]: This's the required implementation of the abstract method inherited from `LLMClientBase`, and the original function signature is:
```py
def _generate(self, system_content: str, user_content: str) -> str:
```
but here it's overridden to:
```py
def _generate(self, prompt: MessageInterface) -> str:
```

Although this still works (since the abstract method `_generate` is indeed implemented), we generally expect subclasses inheriting from an abstract class to have compatible function signatures.

This way, even if we switch to another "implementation," we don’t need to modify how we use the methods.

For example, in the following code, if we replace client with `OtherAIClient()`, we don’t need to update the method call because their function signature are the same:

```py
class BaseClient:
    @abc.abstractmethod
    def method(self, a, b, c):
        ...

class OpenAIClient(BaseClient):
    def method(self, a, b, c):
        ...

class OtherAIClient(BaseClient):
    def method(self, a, b, c):
        ...

client = OpenAIClient() # or OtherAIClient()
client.method(a, b, c) # No need to change method usage even when switching to OtherAIClient()
```

Therefore, it is recommended to use a unified definition.

---

By the way, this function first checks if OpenAI is enabled in your configuration, or returns `""` otherwise. In my opinion, returning an empty string may confuse the caller, and this client should not be initialized if `OpenAI` is disabled in the configuration. I'd recommend moving the check to the initialization phase and raising an exception if necessary.

File: backend/src/llm_client/openai_client.py [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+        Evaluate the relevance of the news title with the prompt.
+        :param news_title:
+        :param prompt:
+        :return:
+        """
+        response = self._generate(MessageInterface(
+            system_content=f"你是一個關聯度評估機器人，請評估新聞標題是否與「{prompt}」相關，並給予'high'、'medium'、'low'評價。(僅需回答'high'、'medium'、'low'三個詞之一)",
+            user_content=news_title
+        ))
+        return response
```
Comment (mlcp29th-hi) [inline backend/src/llm_client/openai_client.py:70]: These are some public methods of `LLMClient`:
1. `extract_search_keywords`
2. `generate_summary`
3. `evaluate_relevance`

All of them create a `MessageInterface` object with system content and user content, and then pass it to `_generate`. It seems to me that we can just directly pass `system_content` and `user_content` to `_generate` so that the statement creating `MessageInterface object can be written just once.

Moreover, `MessageInterface` seems to provide the functionality converting the format of message, given that:
1. The format of message is simple.
2. Each type of client may use different message format.
3. The conversion only be performed at the single place inside this class.

We may remove `MessagePassingInterface`:

```py
def _generate(self, system_content: str, user_content: str):
    completion = self.client.chat.completions.create(
        model=ai_config.OPEN_AI_MODEL,
        messages=[
            {"role": "system", "content": f"{system_content}"},
            {"role": "user", "content": f"{user_content}"}
        ]
    )
    return completion.choices[0].message.content
```
---

Additionally, `evaluate_relevance`, `generate_summary`, and `extract_search_keywords` are implemented directly in the `LLMClient`, and I think there's no issue with that. However, considering the TA's requirement that if a new LLM Client is introduced in the future, it should only need to inherit from `LLMClientBase` and provide implementations for abstract methods, I think these 3 methods could be placed directly in LLMClientBase. Then, subclasses like `LLMClient` only need to implement `_generate`, avoiding the need to rewrite `evaluate_relevance`, `generate_summary`, and `extract_search_keywords` every time a new LLM Client is added.


```py
class LLMClientBase:
    @abc.abstractmethod
    def _generate_text(self, system_content: str, user_content: str):
        raise ImplementedError

    def evaluate_relevance(self, title):
        return self._generate_text(SystemPrompt.EVALUATE_RELEVANCE, title)

    def generate_summary(self, content):
        return self._generate_text(SystemPrompt.GENERATE_SUMMARY, content)

    def extract_search_keywords(self, content):
        return self._generate_text(SystemPrompt.EXTRACT_SEARCH_KEYWORD, content)


class LLMClient(LLMClientBase):
    def _generate_text(self, system_content: str, user_content: str):
        ...

class OtherAIClient(LLMClientBase):
    def _generate_text(self, system_content: str, user_content: str):
        ...

client = LLMClient() # or OtherAIClient()
client.extract_search_keyword("...")
```

File: backend/src/news/router.py
Comment (mlcp29th-hi) [inline backend/src/news/router.py:1]: The changes in this file appear to be mainly replacing the original interaction with `OpenAI` with `LLMClient`. There are no major issues, but the business logic is written directly inside the endpoint functions. It would be better to move it to `service.py` to avoid having both `service.py` and `router.py` each import and instantiate `LLMClient` separately.

File: backend/src/news/service.py
Comment (mlcp29th-hi) [inline backend/src/news/service.py:1]: The module is well refactored to migrate to `LLMClient`. Some of the changes are not related to the goal of this pull request (llm client), and it would be better to include them in the other pull request, which can help the process of code review.

**Review by mlcp29th-hi — COMMENTED** [review]
Hello 👋,
I've reviewed this PR, and my questions and personal suggestions are as follows.

**Comment by Yuannzh1** [comment]
- 靈活的配置管理：
使用 AIConfig 類集中管理配置參數，並支持從 .env 文件或環境變量中動態加載，讓代碼適配不同的部署環境（如開發、測試、和生產環境）更加靈活。

- 性能優化：
利用 lru_cache 緩存配置數據，避免多次重複讀取 .env 文件，減少運行時的性能損耗，提升效率。

- 配置集中化：
所有與 OpenAI API 有關的參數（如模型名稱、API 密鑰等）統一管理，降低配置分散風險，讓開發和維護更加直觀。

- 默認值支持：
提供合理的默認配置（如 OPENAI_ENABLED=True 和 OPENAI_MODEL="gpt-3.5-turbo"），即使環境變量缺失也能保持代碼的穩定運行。

- 易於擴展：
配置管理的結構設計支持新增參數（如超時設置、重試次數等），無需對核心代碼進行大規模改動，適合需求變化的情境。

- 結構簡潔清晰：
通過統一的配置接口（如 get_ai_config()），減少重複代碼，讓代碼邏輯更清晰且易於調用。

結論：通過 AIConfig 集中管理配置參數，提升了代碼的靈活性與性能表現，尤其在多環境部署場景下表現出色。

**Comment by Steven0811** [comment]
對於LLMClient的實現基本上沒甚麼太大的問題，也都有符合軟體工程的原則，對於異常處理的部分也都有做到
在config統一管理模組的關鍵參數也是很好的做法
不過在LLMClientBase中定義的抽象方法稍顯不足，或許可以對LLMClient中的方法有更多的定義或說明

File: backend/src/llm_client/base.py (hunk 2 above)
Comment (patrickhicetc) [inline backend/src/llm_client/base.py:18]: 第15 16行應該沒有用f字串的必要，直接指定就好

File: backend/src/llm_client/openai_client.py (hunk 6 above)
Comment (patrickhicetc) [inline backend/src/llm_client/openai_client.py:70]: 那些很長的字串直接放在這檔案裡看起來有點醜，建議可以再開個檔案存

**Comment by yozen0405** [comment]
倉鼠同學的實作非常細緻，展現了對設計模式和實際需求的深刻理解。特別是 LLMClientBase 的抽象設計，為未來的擴展打下了堅實的基礎，而 MessageInterface 的結構清晰且功能完整，完美封裝了系統與用戶的交互內容，提升了代碼的可讀性與靈活性。同時，對錯誤處理的考慮也非常周到，像是在初始化和生成過程中捕捉具體的 API 異常，確保系統的穩定性和容錯能力。整體來說，這份代碼充分體現了專業水準和用心程度，值得高度讚賞！！👏

**Comment by percykugood** [comment]
整倘上來說沒有太大的問題，但是還是有些建議 :
* 可以進一步抽象出共用的邏輯，例如請求和回應的格式處理，以利未來擴展到其他模型。
* 增強對 API 呼叫失敗的處理機制，例如 API 超時、網路異常或無效回應。

**Comment by EdwardKYC** [comment]
system_content那堆文字是要不要搬到其他地方去 那跟magic number有什麼區別
為何那麼多的change是關於之前的crawler阿 上上週是不是沒改過去這次才調整?
你是不是還有一些conflic沒merge

**Comment by Rogue8464** [comment]
LLMClient 的實現符合設計原則，尤其是在錯誤處理方面，包含了各種可能的錯誤情境，並確保系統的穩定性和容錯能力。並透過統一的config模組來集中管理關鍵參數，建議可以對 LLMClient 中的方法進行更多的定義或說明，以增強清晰度和可擴展性。

File: = convert_news_to_dict(udn_crawler.parse(news.url))
Comment (dee9c-git) [inline = convert_news_to_dict(udn_crawler.parse(news.url)):1]: My brain isn't braining here.

File: backend/src/news/service.py [hunk 7]
```diff
@@ -11,26 +11,34 @@
 from sqlalchemy import delete, insert, select
 from sqlalchemy.orm import Session
 from src.news.config import get_NewsSettings
+from src.crawler.udn_crawler import UDNCrawler
+from src.llm_client.openai_client import LLMClient
+llm_client=LLMClient()
+udn_crawler = UDNCrawler()
```
Comment (dee9c-git) [inline backend/src/news/service.py:17]: Nitpicking: standardize spacing?

**Review by New9c — COMMENTED** [review]
Except for files that I don't understand, the rest does a good job in terms of the main principles :smiley:

**Comment by Yi-Cheng-Wang** [comment]
### Summary  
This code implements an OpenAI-based LLM client, utilizing Pydantic for data model definition and defining the structure of the LLM client through an abstract base class. The code is organized into multiple modules, including configuration, exception handling, and core LLM client logic. This modular design enhances maintainability and scalability.  

### Advantages  
1. **Clear Structure**: The code is divided into multiple modules, making it easier to manage and maintain. Each module has a clear responsibility, adhering to the Single Responsibility Principle.  
2. **Data Validation with Pydantic**: By leveraging Pydantic's `BaseModel`, data validation and serialization are simplified, reducing the need for manual validation.  
3. **Use of Abstract Base Classes**: The abstract base class `LLMClientBase` defines a clear interface, allowing for seamless extension of different LLM client implementations, improving flexibility and scalability.  

### Disadvantages  
1. **Insufficient Error Handling**: In certain scenarios, error handling might be incomplete. For instance, in the `_generate` method, API errors are only logged as messages, without further recovery mechanisms.  
2. **Lack of Logging**: Current error handling relies solely on printing to the console, lacking a systematic logging mechanism. This can hinder troubleshooting and debugging efforts.  
3. **Configuration Management Flexibility**: While environment variables are used for configuration, the approach may lack flexibility for managing different environments. Future improvements could involve adopting a more robust configuration management system.
